_posts/2025-05-05-date-histogram-optimizations.md

```diff
@@ -0,0 +1,216 @@
+
+
+#### Data Slicing
+
+We then tried a different approach - data slicing. Instead of getting the count using aggregation query, we rewrote into normal range query with track all documents to get the documents for single month. This reduced query time dramatically, to about 150 ms for 1 month [9310#issuecomment-1682627380](https://github.com/opensearch-project/OpenSearch/issues/9310#issuecomment-1682627380). Hence, even if we were to do individual operation for each month, the overall query time comes down from 8 seconds to around 2 seconds for whole year without increasing the concurrency and overall cpu cycles.
+
+```json
+{
+  "size": 0,
+  "track_total_hits": "true",
+  "query": {
+    "range": {
+      "dropoff_datetime": {
+        "gte": "2015-01-01 00:00:00",
+        "lt": "2015-02-01 00:00:00"
+      }
+    }
+  }
+}
+```
+
+A major breakthrough came after we spent sometime to understand the execution of this query and why it is much faster than the aggregation query over same time range!!
+
+### Phase 1: Filter Rewrite
+
+Based on our understanding of range query execution, we came up with Filter Rewrite approach. It preemptively created a series of range filters, one for each bucket, in the requested date histogram. For example, the monthly aggregation over year can be rewritten as below:
+
+```json
+{
+  "size": 0,
+  "aggs": {
+    "dropoffs_over_time ": {
+      "filters": {
+        "1420070400000": {
+          "range": {
+            "dropoff_datetime": {
+              "gte": "2015-01-01 00:00:00",
+              "lt": "2015-02-01 00:00:00"
+            }
+          }
+        },
+        "1422748800000": {
+          "range": {
+            "dropoff_datetime": {
+              "gte": "2015-02-01 00:00:00",
+              "lt": "2015-03-01 00:00:00"
+            }
+          }
+        },
+        "1425168000000": {
+          "range": {
+            "dropoff_datetime": {
+              "gte": "2015-03-01 00:00:00",
+              "lt": "2015-04-01 00:00:00"
+            }
+          }
+        },
+        ...
+      }
+    }
+  }
+}
+```
+
+This date histogram aggregation query generates filter corresponding to each bucket and leverages Lucene’s Points Index, which uses BKD trees, to significantly optimize the aggregation. This tree-based structure organizes data into nodes representing value ranges with associated document counts, enabling efficient traversal. By skipping irrelevant subtrees and leveraging early termination, the system reduces unnecessary disk reads and avoids visiting individual documents. which are executed in similar manner as range query using the index tree to determine counts for each bucket faster than document value iteration. Similar approach was also applied to the auto date histogram, composite aggregation on date histogram source, and, later on, numeric range aggregation with consecutive ranges (TODO: Why do we need consecutive ranges??).
```

Comment:
> Why do we need consecutive ranges
It's the outcome of a date histogram.

Reply:
does numeric range aggregation query also need that? IMO, if consecutive ranges is not the requirement of this approach, we need not make that classification

Reply:
Range agg doesn't have such requirement - consecutive ranges. 
The multi-range traversal algorithm should still be able to support in-consecutive ranges case with some modification, like maintain a list of ranges that are active regarding the current travelling point.

Reply:
@bowenlan-amzn - Do you still have any concerns with this?

Reply:
I believe we are on the same page. We don't need to mention `consecutive ranges` as you said.